siflower: riscv64: fix unaligned access for 64bit data

andrewz1 opened the pull request:
This fix need to prevent panic for access to 64bit data not aligned to 64bit address but aligned to 32bit address. This situation possible for example for MHI bus driver, where descriptors (in hardware) looks like:

```C
struct mhi_event_ctxt {
	__le32 intmod;
	__le32 ertype;
	__le32 msivec;

	__le64 rbase __packed __aligned(4);
	__le64 rlen __packed __aligned(4);
	__le64 rp __packed __aligned(4);
	__le64 wp __packed __aligned(4);
};
```

64bit access to 64bit fields of this structures gives kernel panic with reason (5) for read and (7) for write ops. 32bit access to this fields is OK.

This was tested by next code:

```C
#include <stdio.h>
#include <stdlib.h>
#include <stdint.h>
#include <signal.h>
#include <fcntl.h>
#include <sys/mman.h>
#include <unistd.h>

static const char *resource_path = "/sys/bus/pci/devices/0001:00:00.0/resource0"; static int failed = 0;

static void sig_hdl(int s) {
    failed++;
}

int main() {
    int fd, rv = 1;
    size_t page_size = sysconf(_SC_PAGESIZE);
    void *base;
    uint64_t *tstp, tstv;

    if (signal(SIGBUS, sig_hdl) == SIG_ERR) {
        perror("signal");
        return rv;
    }

    if ((fd = open(resource_path, O_RDWR | O_SYNC)) < 0) {
        perror("open");
        return rv;
    }

    base = mmap(NULL, page_size, PROT_READ | PROT_WRITE, MAP_SHARED, fd, 0);
    if (base == MAP_FAILED) {
        perror("mmap");
        goto out_close;
    }

    tstp = base + 4;
    tstv = tstp[0];
    printf("0x%llx\n", tstv);

    munmap(base, page_size);

    if (!failed) {
        printf("pass\n");
        rv = 0;
    } else {
        printf("fail\n");
    }

out_close:
    close(fd);
    return rv;
}
```

andrewz1 committed:
siflower: riscv64: fix unaligned access for 64bit data

This fix need to prevent panic for access to 64bit data not aligned to 64bit
address but aligned to 32bit address. This situation possible for example for
MHI bus driver, where descriptors (in hardware) looks like:

struct mhi_event_ctxt {
	__le32 intmod;
	__le32 ertype;
	__le32 msivec;

	__le64 rbase __packed __aligned(4);
	__le64 rlen __packed __aligned(4);
	__le64 rp __packed __aligned(4);
	__le64 wp __packed __aligned(4);
};

64bit access to 64bit fields of this structures gives kernel panic with reason
(5) for read and (7) for write ops. 32bit access to this fields is OK.

This was tested by next code:

--- cut --- cut --- cut ---
 #include <stdio.h>
 #include <stdlib.h>
 #include <stdint.h>
 #include <signal.h>
 #include <fcntl.h>
 #include <sys/mman.h>
 #include <unistd.h>

static const char *resource_path = "/sys/bus/pci/devices/0001:00:00.0/resource0";
static int failed = 0;

static void sig_hdl(int s) {
    failed++;
}

int main() {
    int fd, rv = 1;
    size_t page_size = sysconf(_SC_PAGESIZE);
    void *base;
    uint64_t *tstp, tstv;

    if (signal(SIGBUS, sig_hdl) == SIG_ERR) {
        perror("signal");
        return rv;
    }

    if ((fd = open(resource_path, O_RDWR | O_SYNC)) < 0) {
        perror("open");
        return rv;
    }

    base = mmap(NULL, page_size, PROT_READ | PROT_WRITE, MAP_SHARED, fd, 0);
    if (base == MAP_FAILED) {
        perror("mmap");
        goto out_close;
    }

    tstp = base + 4;
    tstv = tstp[0];
    printf("0x%llx\n", tstv);

    munmap(base, page_size);

    if (!failed) {
        printf("pass\n");
        rv = 0;
    } else {
        printf("fail\n");
    }

out_close:
    close(fd);
    return rv;
}
--- cut --- cut --- cut ---

Signed-off-by: Andrey Safonov <andrey.safonov@gmail.com>

LGA1150 commented:
Hi Andrew,

C908 supports hardware unaligned access, but with exceptions:
- Unaligned atomic ops are not supported.
- Unaligned accesses to _strongly-ordered_ memory regions are not supported.

As you mentioned _descriptors_, I think it's the second case. So a better solution is to map the descriptor memory as _weakly-ordered uncacheable_ memory.

LGA1150 commented:
ping @981213

andrewz1 commented:
@LGA1150 I understand, but not sure it's possible map PCI BAR as weakly-ordered uncacheable memory...

LGA1150 commented:
@andrewz1 
Unaligned accesses to register memory regions should be avoided. I don't think your test code makes sense, as it may fail even on ARM64.

LGA1150 commented:
Try the two patches:

```diff
--- a/arch/riscv/include/asm/io.h
+++ b/arch/riscv/include/asm/io.h
@@ -28,6 +28,10 @@
 #ifdef CONFIG_MMU
 #define IO_SPACE_LIMIT		(PCI_IO_SIZE - 1)
 #define PCI_IOBASE		((void __iomem *)PCI_IO_START)
+
+#define ioremap_wc(addr, size)	\
+	ioremap_prot((addr), (size), __pgprot(_PAGE_KERNEL_NC))
+
 #endif /* CONFIG_MMU */
 
 /*
--- a/arch/riscv/include/asm/pgtable.h
+++ b/arch/riscv/include/asm/pgtable.h
@@ -203,6 +203,7 @@ extern struct pt_alloc_ops pt_ops __meminitdata;
 
 #define PAGE_TABLE		__pgprot(_PAGE_TABLE)
 
+#define _PAGE_KERNEL_NC ((_PAGE_KERNEL & ~_PAGE_MTMASK) | _PAGE_NOCACHE)
 #define _PAGE_IOREMAP	((_PAGE_KERNEL & ~_PAGE_MTMASK) | _PAGE_IO)
 #define PAGE_KERNEL_IO		__pgprot(_PAGE_IOREMAP)
 
```

```diff
--- a/arch/riscv/include/asm/pgtable.h
+++ b/arch/riscv/include/asm/pgtable.h
@@ -652,6 +652,7 @@ static inline pgprot_t pgprot_writecombine(pgprot_t _prot)
 
 	return __pgprot(prot);
 }
+#define pgprot_dmacoherent pgprot_writecombine
 
 /*
  * THP functions
```

andrewz1 commented:
@LGA1150 "unaligned access" caused by drivers 5+ years old from upstream kernel... hardware with this "wrong" descriptors designed by Qualcomm and works fine on other platforms including ARM64... this 64bit registers in any case accessed using 32bit PCI and aligned to 32bit - in any case two transactions. If U know how to fix this on RISC-V - tell me how... Yes, may be this variant not optimal, but it works.

LGA1150 commented:
Is the PCI space mapped by `ioremap_wc`? If so my patches should work.

andrewz1 commented:
@LGA1150 thanks!

LGA1150 commented:
The `ioremap_wc` patch is backported from upstream commit [5ec6b493183c4706216347853d3446c36ed399d1](https://git.kernel.org/pub/scm/linux/kernel/git/next/linux-next.git/commit/?id=5ec6b493183c4706216347853d3446c36ed399d1) and should go to `target/linux/generic/backports-6.{6.12}`.   The `pgprot_dmacoherent` patch can go to the platform-specific patch folder.

andrewz1 commented:
> The `ioremap_wc` patch is backported from upstream commit [5ec6b493183c4706216347853d3446c36ed399d1](https://git.kernel.org/pub/scm/linux/kernel/git/next/linux-next.git/commit/?id=5ec6b493183c4706216347853d3446c36ed399d1) and should go to `target/linux/generic/backport-6.{6,12}`. The `pgprot_dmacoherent` patch can go to the platform-specific patch folder.

fixed, but I place commit [5ec6b493183c4706216347853d3446c36ed399d1](https://git.kernel.org/pub/scm/linux/kernel/git/next/linux-next.git/commit/?id=5ec6b493183c4706216347853d3446c36ed399d1) to target/linux/generic/pending-6.{6,12} because this patch is not in mainline kernel. also fixed `__pgprot(_PAGE_KERNEL_NC)` for 6.6 - in this version here must be `pgprot_val`

LGA1150 commented:
> fixed, but I place commit [5ec6b493183c4706216347853d3446c36ed399d1](https://git.kernel.org/pub/scm/linux/kernel/git/next/linux-next.git/commit/?id=5ec6b493183c4706216347853d3446c36ed399d1) to target/linux/generic/pending-6.{6,12} because this patch is not in mainline kernel.

The `ioremap_wc` patch [is going to v6.17](https://lore.kernel.org/linux-riscv/e754a5b2-ca45-449a-bddd-f11a35058f48@ghiti.fr/). Please do move it into `backport` directory.

LGA1150 commented:
Any updates?

andrewz1 commented:
> Any updates?

Hi @LGA1150!

I just want to split this request to two, one for upstream kernel patches and another for `pgprot_dmacoherent pgprot_writecombine` arch-agnostic. But not free time now :(
